plugin/plugin.go

```diff
@@ -317,6 +320,47 @@ func wrap_write_callback(ds *C.data_set_t, cvl *C.value_list_t, ud *C.user_data_
 	return 0
 }
 
+// First declare some types, interfaces, general functions
+
+// Shutters are objects that when called will shut down the plugin gracefully
+type Shutter interface {
+	Shutdown() error
+}
+
+// shutdownFuncs holds references to all shutdown callbacks
+var shutdownFuncs = make(map[string]Shutter)
+
+//export wrap_shutdown_callback
+func wrap_shutdown_callback() C.int {
+	if len(shutdownFuncs) <= 0 {
+		return -1
+	}
+	for n, s := range shutdownFuncs {
+		if err := s.Shutdown(); err != nil {
+			Errorf("The plugin named %s failed with error %v when called to shutdown", n, s)
+			return -1
+		}
+		break
+	}
+	return 0
+}
+
+// RegisterShutdown registers a shutdown function with the daemon which is called
+// when the plugin is required to shutdown gracefully.
+func RegisterShutdown(name string, s Shutter) error {
+	cName := C.CString(name)
+	cCallback := C.plugin_shutdown_cb(C.wrap_shutdown_callback)
+
+	status, err := C.register_shutdown_wrapper(cName, cCallback)
```

Comment:
The shutdown callback is a bit different from "read" and "write" because you don't get any context when called, i.e. you can't look up the "right" Go function and call it. That means that we should call `register_shutdown_wrapper()` only *once* to register `wrap_shutdown_callback()`. `wrap_shutdown_callback()` should then iterate over `shutdownFuncs` (as it does) and call *all* of them.

In other words, have a global variable that tracks if we called `register_shutdown_wrapper` already. If it has already been called, skip that step and only add the additional callback to `shutdownFuncs` and return.